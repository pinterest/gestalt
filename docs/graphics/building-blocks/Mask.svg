<svg xmlns="http://www.w3.org/2000/svg" width="312" height="180" fill="none"><path fill="#767676" fill-opacity=".5" fill-rule="evenodd" d="M226 20H86v140h140V20Zm-70 120c27.614 0 50-22.386 50-50s-22.386-50-50-50-50 22.386-50 50 22.386 50 50 50Z" clip-rule="evenodd"/><path fill="#4A4A4A" fill-rule="evenodd" d="M85 20a1 1 0 0 1 1-1h3.889a1 1 0 1 1 0 2H87v2.889a1 1 0 1 1-2 0V20Zm11.667 0a1 1 0 0 1 1-1h7.777a1 1 0 1 1 0 2h-7.777a1 1 0 0 1-1-1Zm15.555 0a1 1 0 0 1 1-1H121a1 1 0 0 1 0 2h-7.778a1 1 0 0 1-1-1Zm15.556 0a1 1 0 0 1 1-1h7.778a1 1 0 0 1 0 2h-7.778a1 1 0 0 1-1-1Zm15.555 0a1 1 0 0 1 1-1h7.778a1 1 0 0 1 0 2h-7.778a1 1 0 0 1-1-1Zm15.556 0a1 1 0 0 1 1-1h7.778a1 1 0 0 1 0 2h-7.778a1 1 0 0 1-1-1Zm15.555 0a1 1 0 0 1 1-1h7.778a1 1 0 1 1 0 2h-7.778a1 1 0 0 1-1-1ZM190 20a1 1 0 0 1 1-1h7.778a1 1 0 0 1 0 2H191a1 1 0 0 1-1-1Zm15.556 0a1 1 0 0 1 1-1h7.777a1 1 0 1 1 0 2h-7.777a1 1 0 0 1-1-1Zm15.555 0a1 1 0 0 1 1-1H226a1 1 0 0 1 1 1v3.889a1 1 0 0 1-2 0V21h-2.889a1 1 0 0 1-1-1ZM226 30.667a1 1 0 0 1 1 1v7.777a1 1 0 0 1-2 0v-7.777a1 1 0 0 1 1-1Zm-140 0a1 1 0 0 1 1 1v7.777a1 1 0 1 1-2 0v-7.777a1 1 0 0 1 1-1Zm65.996 8.488a51.715 51.715 0 0 1 8.008 0 1 1 0 1 1-.155 1.994 49.707 49.707 0 0 0-7.698 0 1 1 0 1 1-.155-1.994Zm-6.701 1.983a1 1 0 0 1-.74 1.206 48.59 48.59 0 0 0-7.311 2.374 1 1 0 1 1-.766-1.848 50.6 50.6 0 0 1 7.612-2.472 1 1 0 0 1 1.205.74Zm21.41 0a1 1 0 0 1 1.205-.74 50.6 50.6 0 0 1 7.612 2.472 1 1 0 1 1-.766 1.848 48.59 48.59 0 0 0-7.311-2.375 1 1 0 0 1-.74-1.205ZM226 46.222a1 1 0 0 1 1 1V55a1 1 0 0 1-2 0v-7.778a1 1 0 0 1 1-1Zm-140 0a1 1 0 0 1 1 1V55a1 1 0 1 1-2 0v-7.778a1 1 0 0 1 1-1Zm44.724.616a1 1 0 0 1-.329 1.376 49.221 49.221 0 0 0-6.218 4.525 1 1 0 0 1-1.3-1.52 51.183 51.183 0 0 1 6.472-4.71 1 1 0 0 1 1.375.33Zm50.552 0a1 1 0 0 1 1.375-.329 51.183 51.183 0 0 1 6.472 4.71 1 1 0 0 1-1.3 1.52 49.221 49.221 0 0 0-6.218-4.525 1 1 0 0 1-.329-1.376Zm-62.647 9.93a1 1 0 0 1 .11 1.409 49.146 49.146 0 0 0-4.525 6.218 1 1 0 0 1-1.705-1.046 51.216 51.216 0 0 1 4.71-6.472 1 1 0 0 1 1.41-.11Zm74.742 0a1 1 0 0 1 1.41.11 51.216 51.216 0 0 1 4.71 6.47 1 1 0 0 1-1.705 1.047 49.146 49.146 0 0 0-4.525-6.218 1 1 0 0 1 .11-1.41ZM226 61.777a1 1 0 0 1 1 1v7.777a1 1 0 0 1-2 0v-7.777a1 1 0 0 1 1-1Zm-140 0a1 1 0 0 1 1 1v7.778a1 1 0 0 1-2 0v-7.778a1 1 0 0 1 1-1Zm24.177 8.159a1 1 0 0 1 .541 1.307 48.638 48.638 0 0 0-2.375 7.311 1 1 0 1 1-1.945-.465 50.78 50.78 0 0 1 2.472-7.612 1 1 0 0 1 1.307-.541Zm91.646 0a1 1 0 0 1 1.307.54 50.78 50.78 0 0 1 2.472 7.613 1 1 0 0 1-1.945.465 48.638 48.638 0 0 0-2.375-7.311 1 1 0 0 1 .541-1.307ZM226 77.333a1 1 0 0 1 1 1v7.778a1 1 0 0 1-2 0v-7.778a1 1 0 0 1 1-1Zm-140 0a1 1 0 0 1 1 1v7.778a1 1 0 1 1-2 0v-7.778a1 1 0 0 1 1-1Zm119.771 7.743a1 1 0 0 1 1.074.92 51.525 51.525 0 0 1 0 8.008 1 1 0 0 1-1.994-.155 49.524 49.524 0 0 0 0-7.698 1 1 0 0 1 .92-1.075Zm-99.542 0a1 1 0 0 1 .92 1.075 49.524 49.524 0 0 0 0 7.698 1 1 0 1 1-1.994.155 51.525 51.525 0 0 1 0-8.008 1 1 0 0 1 1.074-.92ZM226 92.89a1 1 0 0 1 1 1v7.778a1 1 0 0 1-2 0v-7.778a1 1 0 0 1 1-1Zm-140 0a1 1 0 0 1 1 1v7.778a1 1 0 0 1-2 0v-7.778a1 1 0 0 1 1-1Zm21.138 7.816a1 1 0 0 1 1.205.74 48.625 48.625 0 0 0 2.375 7.311 1 1 0 0 1-1.848.766 50.772 50.772 0 0 1-2.472-7.612 1 1 0 0 1 .74-1.205Zm97.724 0a1 1 0 0 1 .74 1.205 50.772 50.772 0 0 1-2.472 7.612 1 1 0 1 1-1.848-.766 48.625 48.625 0 0 0 2.375-7.311 1 1 0 0 1 1.205-.74ZM226 108.444a1 1 0 0 1 1 1v7.778a1 1 0 1 1-2 0v-7.778a1 1 0 0 1 1-1Zm-140 0a1 1 0 0 1 1 1v7.778a1 1 0 1 1-2 0v-7.778a1 1 0 0 1 1-1Zm26.838 6.832a1 1 0 0 1 1.376.329 49.2 49.2 0 0 0 4.525 6.218 1 1 0 0 1-1.52 1.3 51.238 51.238 0 0 1-4.71-6.472 1 1 0 0 1 .329-1.375Zm86.324 0a1 1 0 0 1 .329 1.375 51.238 51.238 0 0 1-4.71 6.472 1 1 0 0 1-1.52-1.3 49.2 49.2 0 0 0 4.525-6.218 1 1 0 0 1 1.376-.329ZM226 124a1 1 0 0 1 1 1v7.778a1 1 0 0 1-2 0V125a1 1 0 0 1 1-1Zm-140 0a1 1 0 0 1 1 1v7.778a1 1 0 0 1-2 0V125a1 1 0 0 1 1-1Zm36.767 3.371a1 1 0 0 1 1.41-.11 49.2 49.2 0 0 0 6.218 4.525 1 1 0 0 1-1.046 1.705 51.238 51.238 0 0 1-6.472-4.71 1 1 0 0 1-.11-1.41Zm66.466 0a1 1 0 0 1-.11 1.41 51.238 51.238 0 0 1-6.472 4.71 1 1 0 0 1-1.046-1.705 49.2 49.2 0 0 0 6.218-4.525 1 1 0 0 1 1.41.11Zm-53.296 8.452a1 1 0 0 1 1.307-.541 48.625 48.625 0 0 0 7.311 2.375 1 1 0 1 1-.465 1.945 50.772 50.772 0 0 1-7.612-2.472 1 1 0 0 1-.541-1.307Zm40.126 0a1 1 0 0 1-.541 1.307 50.772 50.772 0 0 1-7.612 2.472 1 1 0 0 1-.465-1.945 48.625 48.625 0 0 0 7.311-2.375 1 1 0 0 1 1.307.541Zm-24.987 3.948a1 1 0 0 1 1.075-.92 49.508 49.508 0 0 0 7.698 0 1 1 0 1 1 .155 1.994 51.516 51.516 0 0 1-8.008 0 1 1 0 0 1-.92-1.074Zm74.924-.215a1 1 0 0 1 1 1v7.777a1 1 0 1 1-2 0v-7.777a1 1 0 0 1 1-1Zm-140 0a1 1 0 0 1 1 1v7.777a1 1 0 1 1-2 0v-7.777a1 1 0 0 1 1-1Zm0 15.555a1 1 0 0 1 1 1V159h2.889a1 1 0 0 1 0 2H86a1 1 0 0 1-1-1v-3.889a1 1 0 0 1 1-1Zm140 0a1 1 0 0 1 1 1V160a1 1 0 0 1-1 1h-3.889a1 1 0 0 1 0-2H225v-2.889a1 1 0 0 1 1-1ZM96.667 160a1 1 0 0 1 1-1h7.777a1 1 0 1 1 0 2h-7.777a1 1 0 0 1-1-1Zm15.555 0a1 1 0 0 1 1-1H121a1 1 0 0 1 0 2h-7.778a1 1 0 0 1-1-1Zm15.556 0a1 1 0 0 1 1-1h7.778a1 1 0 0 1 0 2h-7.778a1 1 0 0 1-1-1Zm15.555 0a1 1 0 0 1 1-1h7.778a1 1 0 0 1 0 2h-7.778a1 1 0 0 1-1-1Zm15.556 0a1 1 0 0 1 1-1h7.778a1 1 0 0 1 0 2h-7.778a1 1 0 0 1-1-1Zm15.555 0a1 1 0 0 1 1-1h7.778a1 1 0 1 1 0 2h-7.778a1 1 0 0 1-1-1ZM190 160a1 1 0 0 1 1-1h7.778a1 1 0 0 1 0 2H191a1 1 0 0 1-1-1Zm15.556 0a1 1 0 0 1 1-1h7.777a1 1 0 1 1 0 2h-7.777a1 1 0 0 1-1-1Z" clip-rule="evenodd"/></svg>