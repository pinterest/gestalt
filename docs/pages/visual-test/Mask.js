// @flow strict
import { type Node as ReactNode } from 'react';
import { Box, ColorSchemeProvider, Flex, Image, Mask } from 'gestalt';

export default function Snapshot(): ReactNode {
  return (
    <ColorSchemeProvider colorScheme="light">
      <Box color="default" display="inlineBlock" maxWidth={308} padding={1}>
        <Flex maxWidth={300} wrap>
          {[0, 2, 4, 6, 8, 'circle'].map((a) => (
            <Box key={a} maxWidth={100}>
              <Mask rounding={a}>
                <Box height={100} width={100}>
                  <Image
                    alt="tall"
                    color="black"
                    naturalHeight={1}
                    naturalWidth={1}
                    src="https://d3cy9zhslanhfa.cloudfront.net/media/BBEEEEC7-E954-4223-B5A061E37D0C03E2/CE43CF95-DE36-465B-956EFB21C9CC9C04/webimage-0311D236-89DC-4404-9D9B1452C865159C.png"
                  />
                </Box>
              </Mask>
            </Box>
          ))}
        </Flex>
      </Box>
    </ColorSchemeProvider>
  );
}
