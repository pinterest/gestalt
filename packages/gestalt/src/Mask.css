.Mask {
  composes: overflowHidden from "./Layout.css";
  composes: relative from "./Layout.css";
}

.willChangeTransform {
  /* â¬‡ https://github.com/pinterest/gestalt/pull/965 */
  will-change: transform;
}

.square {
  composes: square from "./Borders.css";
}

.rounded {
  composes: rounded from "./Borders.css";
}

.circle {
  composes: circle from "./Borders.css";
}

.wash {
  composes: absolute from "./Layout.css";
  composes: bottom0 from "./Layout.css";
  composes: left0 from "./Layout.css";
  composes: right0 from "./Layout.css";
  composes: top0 from "./Layout.css";
  background: rgba(0, 0, 0, 0.03);
  pointer-events: none;
}
