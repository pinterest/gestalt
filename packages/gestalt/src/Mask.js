// @flow
import * as React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import styles from './Mask.css';

type Props = {|
  children?: React.Node,
  height?: number | string,
  shape?: 'circle' | 'rounded' | 'square',
  width?: number | string,
  willChangeTransform?: boolean,
  wash?: boolean,
|};

export default function Mask(props: Props) {
  const {
    children,
    shape = 'square',
    width,
    height,
    willChangeTransform = true,
    wash = false,
  } = props;
  return (
    <div
      className={cx(styles.Mask, styles[shape], {
        [styles.willChangeTransform]: willChangeTransform,
      })}
      style={{ width, height }}
    >
      {children}
      {wash && <div className={styles.wash} />}
    </div>
  );
}

Mask.propTypes = {
  children: PropTypes.node,
  height: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  shape: PropTypes.oneOf(['circle', 'rounded', 'square']),
  width: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  willChangeTransform: PropTypes.bool,
  wash: PropTypes.bool,
};
