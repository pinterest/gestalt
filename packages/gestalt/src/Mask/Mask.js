// @flow
import * as React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import styles from './Mask.css';

type Props = {|
  children?: React.Node,
  height?: number | string,
  shape?: 'circle' | 'rounded' | 'square',
  width?: number | string,
  wash?: boolean,
|};

export default function Mask(props: Props) {
  const { children, shape = 'square', width, height, wash = false } = props;
  return (
    <div className={cx(styles.Mask, styles[shape])} style={{ width, height }}>
      {children}
      {wash && <div className={styles.wash} />}
    </div>
  );
}

Mask.propTypes = {
  children: PropTypes.node,
  height: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  shape: PropTypes.oneOf(['circle', 'rounded', 'square']),
  width: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  wash: PropTypes.bool,
};
