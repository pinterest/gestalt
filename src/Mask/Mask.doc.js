// @flow
import * as React from 'react';
import Mask from './Mask';
import {
  ns,
  card,
  md,
  PropTable,
  Example,
  Combination,
} from '../../docs/src/cards';

ns(
  'Mask',
  `If you have an item you need to fit into a shape, you can achieve this by
  putting a \`Mask\` on it.
  `
);

card(
  <PropTable
    props={[
      {
        name: 'children',
        type: 'any',
      },
      {
        name: 'height',
        type: `number | string`,
      },
      {
        name: 'width',
        type: `number | string`,
      },
      {
        name: 'shape',
        type: `"circle" | "rounded" | "square"`,
        defaultValue: 'square',
      },
      {
        name: 'wash',
        type: 'boolean',
        defaultValue: false,
      },
    ]}
  />,
  { heading: false }
);

card(
  'Example',
  <Example
    defaultCode={`
<Mask height={70} shape="circle" width={70}>
  <div style={{ backgroundColor: '#0fa573', width: 70, height: 70 }} />
</Mask>
`}
    scope={{ Mask }}
  />,
  { stacked: true }
);

card(
  'Example: Masking other content',
  md`
    You can compose images with other content (like images or videos) to produce different shapes like rounded rectangles or circles.
  `,
  <Example
    defaultCode={`
<Box maxWidth={300}>
  <Mask shape="circle">
    <img
      alt="weakendclub.com"
      src="https://s-media-cache-ak0.pinimg.com/564x/4f/cb/61/4fcb610ae43cddd68086c85eec9c413e.jpg"
      style={{ maxWidth: '100%', display: 'block' }}
    />
  </Mask>
</Box>
`}
    scope={{ Mask }}
  />,
  { stacked: true }
);

card(
  'Example: Adding a wash',
  md`
    If you expect the masked content to be nearly white, you can apply a wash to emphasize the edge of the mask.
  `,
  <Example
    defaultCode={`
<Box maxWidth={300}>
  <Mask shape="rounded" wash>
    <img
      alt="subliming.tumblr.com"
      src="https://s-media-cache-ak0.pinimg.com/564x/d0/c1/f6/d0c1f69316b4e61b3ccc7b1731f028b8.jpg"
      style={{ maxWidth: '100%', display: 'block' }}
    />
  </Mask>
</Box>
`}
    scope={{ Mask }}
  />,
  { stacked: true }
);

card(
  'Shape Combinations',
  <Combination shape={['circle', 'rounded', 'square']}>
    {props => (
      <Mask height={70} width={70} {...props}>
        <div style={{ backgroundColor: '#e3780c', width: 70, height: 70 }} />
      </Mask>
    )}
  </Combination>
);
