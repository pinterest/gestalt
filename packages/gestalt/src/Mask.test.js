// @flow
import React from 'react';
import renderer from 'react-test-renderer';
import Mask from './Mask.js';

test('Mask renders', () => {
  const component = renderer.create(
    <Mask>
      <img alt="Painting" src="https://pinterest.com/img/painting.png" />
    </Mask>
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Mask has the correct width', () => {
  const component = renderer.create(
    <Mask width={400}>
      <img alt="Painting" src="https://pinterest.com/img/painting.png" />
    </Mask>
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
