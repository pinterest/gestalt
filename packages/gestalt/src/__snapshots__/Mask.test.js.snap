// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mask has the correct width 1`] = `
<div
  className="Mask square willChangeTransform"
  style={
    Object {
      "height": undefined,
      "width": 400,
    }
  }
>
  <img
    alt="Painting"
    src="https://pinterest.com/img/painting.png"
  />
</div>
`;

exports[`Mask renders 1`] = `
<div
  className="Mask square willChangeTransform"
  style={
    Object {
      "height": undefined,
      "width": undefined,
    }
  }
>
  <img
    alt="Painting"
    src="https://pinterest.com/img/painting.png"
  />
</div>
`;
